api/xds_protocol.rst

```diff
@@ -155,45 +266,116 @@ the resources in the :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`
+<envoy_api_msg_DiscoveryRequest>`; in the incremental protocol variants, this is done via the
+:ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>` and
+:ref:`resource_names_unsubscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_unsubscribe>` fields in the
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`.
+
+Normally (see below for exceptions), requests must specify the set of resource names that the client is interested in. The
+management server must supply the requested resources if they exist. The client will silently ignore any supplied resources that
+were not explicitly requested. When the client sends a new request that changes the set of resources being requested, the server
+must resend any newly requested resources, even if it previously sent those resources without having been asked for them and the
+resources have not changed since that time. If the list of resource names becomes empty, that means that the client is no longer
+interested in any resources of the specified type.
+
+For :ref:`Listener <envoy_api_msg_Listener>` and :ref:`Cluster <envoy_api_msg_Cluster>` resource types, there is also a
+"wildcard" mode, which is triggered when the initial request on the stream for that resource type contains no resource names.
+In this case, the server should use site-specific business logic to determine the full set of resources that the client is
+interested in, typically based on the client's :ref:`node <envoy_api_msg_Core.Node>` identification. Note that once a
+stream has entered wildcard mode for a given resource type, there is no way to change the stream out of wildcard mode;
+resource names specified in any subsequent request on the stream will be ignored.
+
+Envoy will always use wildcard mode for :ref:`Listener <envoy_api_msg_Listener>` and :ref:`Cluster <envoy_api_msg_Cluster>`
+resources. However, other xDS clients (such as gRPC clients that use xDS) may specify explicit resource names for these
+resource types, for example if they only have a singleton listener and already know its name from some out-of-band configuration.
+
+Grouping Resources into Responses
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the incremental protocol variants, the server sends each resource in its own response. This means that if the server has
+previously sent 100 resources and only one of them has changed, it may send a response containing only the changed resource; it
+does not need to resend the 99 resources that have not changed, and the client must not delete the unchanged resources.
+
+In the SotW protocol variants, all resource types except for :ref:`Listener <envoy_api_msg_Listener>` and :ref:`Cluster
+<envoy_api_msg_Cluster>` are grouped into responses in the same way as in the incremental protocol variants. However,
+:ref:`Listener <envoy_api_msg_Listener>` and :ref:`Cluster <envoy_api_msg_Cluster>` resource types are handled differently:
```

Comment:
I think this is a bit misleading, since for SoTW EDS, all hosts have to be sent even if one is being added/removed, right? Can you rephrase?

Reply:
All hosts need to be sent, but not all `ClusterLoadAssignment`s, in an EDS update. I.e. if  we have 4 clusters, {A, B, C, D}, we can send one EDS update with two `ClusterLoadAssignment`s for {A, B}, with all the hosts for A and B in the respective messages. Then we can do the same for {C, D}.

Reply:
Right, I would just clarify the text a bit if possible.

Reply:
Can you suggest what wording you'd like to see here?  This document is talking about the set of resource types defined at the top, and individual hosts are not one of those types, so it's not clear to me how this could be misread to assume that individual hosts could be sent separately.  Why would a reader think that hosts are any different from any other repeated field in the API?

Reply:
I think my confusion here was basically related to the discussion starting around here: https://github.com/envoyproxy/envoy/issues/8400#issuecomment-559883863. I had forgotten that we have a nested CLA message and that we are still in a situation in which all hosts have to be sent either in incremental or SoTW. I think this is such a common thing and a confusion point that if you are willing to explicitly call it out that would be really great.

Reply:
Ah, okay.  I've added a note about this.